### ISSUE TITLE
NaN in gradient

### ISSUE THREAD

**Issue by lamthuy** [jcmgray/quimb#355]
### What is your issue?

Hi I create a trainable MPO as follows:

```
import torch
import quimb.tensor as qtn


class NeuralMPO(torch.nn.Module):
    def __init__(self, L, phys_dim=2, bond_dim=2, dtype=torch.float32):
        super().__init__()

        self.L = L
        self.dtype = dtype

        mpo = qtn.MPO_rand(L, bond_dim=bond_dim, phys_dim=phys_dim)
        mpo.apply_to_arrays(lambda x: torch.tensor(x, dtype=dtype))
        params, self.skeleton = qtn.pack(mpo)
        self.torch_params = torch.nn.ParameterDict({
            str(i): torch.nn.Parameter(p)
            for i, p in params.items()
        })

    def forward(self, psi_in, psi_target, max_bond=32):

        # reconstruct MPO
        params = {int(i): p for i, p in self.torch_params.items()}
        W = qtn.unpack(params, self.skeleton)

        # apply MPO to MPS
        psi_pred = W.apply(psi_in)
        psi_pred.compress(max_bond=max_bond)

        # normalize states
        psi_pred.normalize()

        psi_target = psi_target.copy()
        psi_target.normalize()

        # compute overlap
        overlap = psi_pred.H @ psi_target

        # ensure torch tensor
        # overlap = torch.as_tensor(overlap, dtype=self.dtype)

        loss = 1.0 - torch.abs(overlap) ** 2
        return loss
```
The training works well if I comment out the line:
```
# psi_pred.compress(max_bond=max_bond)
```
When I do not do so, in many runs I got NaN in gradients. I think because the SVD process used to compress the MPS may create unstable gradients when I do the backward? Is there any way to resolve this issue?

**Comment by jcmgray** [comment]
Hi @lamthuy, sorry for the slow response! I don't really have a definitive way to resolve this. 

One problem it might be is that degenerate singular values create an undefined gradient. You can find various implementations of stable SVD gradients (that e.g. simply take a safe inverse of the singular value differences), torch then allows you customize the SVD op including gradient. 

If you go down this route it might be nice to have in `quimb` a `enable_safe_svd_gradient_torch` function that switches out the implementation.

Another option might be to explore different compression methods, such as using randomized svd.
